### PR TITLE
Support components for entity and container (Nameable) custom names

### PR DESCRIPTION
**Pull request by jpenilla**
Tried my best to do this correctly. Tested and working.

[Paperclip](https://anonymousfiles.io/DyaQCX2H/) for testing (``java -Dpaperclip.install=true -jar paperclip.jar`` to install to local maven)

### CODE REVIEW

**Comment by imDaniX** [comment]
I guess related to https://github.com/PaperMC/Paper/issues/3715

File: Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
++++ b/src/main/java/org/bukkit/Nameable.java
+@@ -28,4 +28,31 @@ public interface Nameable {
+      * @param name the name to set
+      */
+     public void setCustomName(@Nullable String name);
++
++    // Paper start
++    /**
++     * Gets the custom name on a mob or block. If there is no name this method
++     * will return null.
```
Comment (Proximyst) [inline Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch:19]:
```suggestion
+     * will return {@code null}.
```

File: Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
++     * <p>
++     * This value has no effect on players, they will always use their real
```
Comment (Proximyst) [inline Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch:21]:
```suggestion
+     * This value has no effect on players: they will always use their real
```

File: Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
++     * name.
++     *
++     * @return name of the mob/block or null
```
Comment (Proximyst) [inline Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch:24]:
```suggestion
+     * @return name of the mob/block or {@code null} if none
```

File: Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
++     */
++    @Nullable
++    public net.md_5.bungee.api.chat.BaseComponent[] getCustomNameComponent();
++
++    /**
++     * Sets a custom name on a mob or block. This name will be used in death
++     * messages and can be sent to the client as a nameplate over the mob.
++     * <p>
++     * Setting the name to null will clear it.
```
Comment (Proximyst) [inline Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch:33]:
```suggestion
+     * Setting the name to {@code null} will clear it.
```

File: Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
++     * <p>
++     * This value has no effect on players, they will always use their real
```
Comment (Proximyst) [inline Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch:35]:
```suggestion
+     * This value has no effect on players: they will always use their real
```

File: Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
++     * name.
++     *
++     * @param name the name to set
```
Comment (Proximyst) [inline Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch:38]:
```suggestion
+     * @param name the name to set, or {@code null} to clear it
```

**Review by Proximyst — APPROVED** [review]
LGTM. Minor issues we can resolve on merge.

**Review by zachbr — DISMISSED** [review]
I'm fine with this if Proxi is happy with it. If Proxi wants to wait on Mini that's okay too. More eyes are good.

File: Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+--- a/src/main/java/org/bukkit/Nameable.java
++++ b/src/main/java/org/bukkit/Nameable.java
+@@ -28,4 +28,31 @@ public interface Nameable {
+      * @param name the name to set
+      */
+     public void setCustomName(@Nullable String name);
++
++    // Paper start
++    /**
++     * Gets the custom name on a mob or block. If there is no name this method
```
Comment (MiniDigger) [inline Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch:18]:
```suggestion
+     * Gets the custom name of a mob or block. If there is no name this method
```

File: Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
++     * This value has no effect on players, they will always use their real
++     * name.
++     *
++     * @return name of the mob/block or null
++     */
++    @Nullable
++    public net.md_5.bungee.api.chat.BaseComponent[] getCustomNameComponent();
++
++    /**
++     * Sets a custom name on a mob or block. This name will be used in death
```
Comment (MiniDigger) [inline Spigot-API-Patches/0227-Support-components-in-Nameable-Entites-Blocks.patch:30]:
```suggestion
+     * Sets a custom name of a mob or block. This name will be used in death
```

**Review by MiniDigger — APPROVED** [review]

**Comment by kashike** [comment]
We are handling this another way - tracking in #4830.

**Pull request closed without merging**